http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java

```diff
@@ -57,20 +58,34 @@ public void onResponseHeaders(HttpStream stream, int responseStatusCode, int hea
      * Release the connection back to the pool so that it can be reused.
      */
     public void releaseConnection(HttpStream stream) {
-        if (connectionClosed.compareAndSet(false, true)) {
-            connection.close();
-            stream.close();
+        synchronized (lock) {
+            if (!connectionClosed) {
+                connectionClosed = true;
+                connection.close();
+                stream.close();
+            }
+        }
+    }
+
+    public void incrementWindow(HttpStream stream, int windowSize) {
```

Comment:
Could you please confirm if we have testes and confirmed that  there is not latency because of this synchronized check in multi threaded environment  with multiple requests going on at the same time?

Reply:
I don't anticipate any latency change as part of this change since most of times, there should only be one thread accessing those methods. The only exception is that when the request fails, it may be possible one thread is trying to invoke incrementWindow and the other is closing it.

I'll double check our benchmarks after today's release.